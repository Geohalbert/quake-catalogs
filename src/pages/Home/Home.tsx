import React, { useEffect, memo } from 'react';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';

import { quakesAction } from '../../actions';
import { QuakeList } from '../../components';
import { AppState, ThunkDispatch } from '../../types';
import styles from './styles.scss';

type Props = ReturnType<typeof mapStateToProps> &
  ReturnType<typeof mapDispatchToProps>;

// Export for unit testing
export const Home = ({ readyStatus, list, fetchQuakesIfNeeded }: Props) => {
  useEffect(() => {
    fetchQuakesIfNeeded();
  }, []);

  const renderQuakeList = () => {
    if (!readyStatus || readyStatus === 'invalid' || readyStatus === 'request')
      return <p>Loading...</p>;

    if (readyStatus === 'failure') return <p>Oops, Failed to load list!</p>;

    return <QuakeList list={list} />;
  };

  return (
    <div className={styles.Home}>
      <Helmet title="Home" />
      {renderQuakeList()}
    </div>
  );
};

const mapStateToProps = ({ home: { readyStatus, list } }: AppState) => ({
  readyStatus,
  list
});

const mapDispatchToProps = (dispatch: ThunkDispatch) => ({
  fetchQuakesIfNeeded: () => dispatch(quakesAction.fetchQuakesIfNeeded())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(memo(Home));
