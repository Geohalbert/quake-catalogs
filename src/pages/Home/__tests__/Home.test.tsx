import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import { Home } from '../Home';

describe('<Home />', () => {
  const tree = (props: object, actions: object) =>
    renderer
      .create(
        <MemoryRouter>
          {/*
            // @ts-ignore */}
          <Home {...props} {...actions} />
        </MemoryRouter>
      )
      .toJSON();

  it('should call fetchQuakesIfNeeded when componentDidMount', () => {
    const mockAction = jest.fn();
    const actions = { fetchQuakesIfNeeded: mockAction };

    mount(
      <MemoryRouter>
        {/*
          // @ts-ignore */}
        <Home {...actions} />
      </MemoryRouter>
    );

    expect(mockAction).toHaveBeenCalled();
  });

  it('renders the loading status if data invalid', () => {
    const props = { readyStatus: 'invalid' };
    const actions = { fetchQuakesIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

  it('renders the loading status if requesting data', () => {
    const props = { readyStatus: 'request' };
    const actions = { fetchQuakesIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

  it('renders an error if loading failed', () => {
    const props = { readyStatus: 'failure' };
    const actions = { fetchQuakesIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

  it('renders the <QuakeList /> if loading was successful', () => {
    const props = {
      readyStatus: 'success',
      list: [{ id: '1', name: 'Welly' }]
    };
    const actions = { fetchQuakesIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });
});
