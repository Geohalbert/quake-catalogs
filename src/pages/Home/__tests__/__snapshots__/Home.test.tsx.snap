// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders an error if loading failed 1`] = `
<div
  className="Home"
>
  <p>
    Oops, Failed to load list!
  </p>
</div>
`;

exports[`<Home /> renders the <QuakeList /> if loading was successful 1`] = `
<div
  className="Home"
>
  <div
    className="QuakeList"
  >
    <h4>
      Quake List
    </h4>
    <ul>
      <li>
        <a
          href="/QuakeInfo/1"
          onClick={[Function]}
        >
          Welly
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`<Home /> renders the loading status if data invalid 1`] = `
<div
  className="Home"
>
  <p>
    Loading...
  </p>
</div>
`;

exports[`<Home /> renders the loading status if requesting data 1`] = `
<div
  className="Home"
>
  <p>
    Loading...
  </p>
</div>
`;
